llpc/context/llpcCompiler.cpp

```diff
@@ -1353,85 +1353,66 @@ void GraphicsShaderCacheChecker::UpdateAndMerge(
-            // Load ELF binary
-            auto result = writer.ReadFromBuffer(nonFragmentPipelineElf.pCode, nonFragmentPipelineElf.codeSize);
-            assert(result == Result::Success);
-            (void(result)); // unused
-            writer.MergeElfBinary(m_pContext, &m_fragmentElf, pPipelineElf);
+        pipelineElf.codeSize = pPipelineElf->size();
+        pipelineElf.pCode = pPipelineElf->data();
 
-            pipelineElf.codeSize = pPipelineElf->size();
-            pipelineElf.pCode = pPipelineElf->data();
+        if (m_fragmentCacheEntryState == ShaderEntryState::Compiling)
+        {
+            m_pCompiler->UpdateShaderCache(result == Result::Success, &pipelineElf, m_pFragmentShaderCache,
+                                           m_hFragmentEntry);
         }
 
-        m_pCompiler->UpdateShaderCache(result == Result::Success, &pipelineElf, m_pNonFragmentShaderCache,
-                                       m_hNonFragmentEntry);
+        if (m_nonFragmentCacheEntryState == ShaderEntryState::Compiling)
+        {
+            m_pCompiler->UpdateShaderCache(result == Result::Success, &pipelineElf, m_pNonFragmentShaderCache,
+                                           m_hNonFragmentEntry);
+        }
     }
 
-    // Only fragment shader is compiled
-    else if ((m_nonFragmentCacheEntryState == ShaderEntryState::Ready) &&
-             (m_fragmentCacheEntryState == ShaderEntryState::Compiling))
+    // Now merge ELFs if one or both parts are from the cache. Nothing needs to be merged if we just compiled the full
+    // pipeline, as everything is already contained in the single incoming ELF in this case.
+    if (result == Result::Success &&
+        (m_fragmentCacheEntryState == ShaderEntryState::Ready ||
+         m_nonFragmentCacheEntryState == ShaderEntryState::Ready))
     {
-        ElfPackage partialPipelineElf = *pPipelineElf;
+        // Move the compiled ELF out of the way.
+        ElfPackage compiledPipelineElf = std::move(*pPipelineElf);
         pPipelineElf->clear();
-        BinaryData pipelineElf = {};
-        if (result == Result::Success)
-        {
-            BinaryData fragmentPipelineElf = {};
-            fragmentPipelineElf.pCode = partialPipelineElf.data();
-            fragmentPipelineElf.codeSize = partialPipelineElf.size();
 
-            ElfWriter<Elf64> writer(m_pContext->GetGfxIpVersion());
-            // Load ELF binary
-            auto result = writer.ReadFromBuffer(m_nonFragmentElf.pCode, m_nonFragmentElf.codeSize);
-            assert(result == Result::Success);
-            (void(result)); // unused
-
-            writer.MergeElfBinary(m_pContext, &fragmentPipelineElf, pPipelineElf);
-
-            pipelineElf.codeSize = pPipelineElf->size();
-            pipelineElf.pCode = pPipelineElf->data();
+        // Determine where the fragment / non-fragment parts come from (cache or just-compiled).
+        BinaryData fragmentElf;
```

Comment:
fragmentElf -> fragmentElf = {}
I understand you have it initialized in the following codes. But as suggested in PAL coding doc, it is better to use "= {}" to initialize structures unconditionally because you might have additional fields added in the future and it is easy for you to forget to initialize those new fields.

Reply:
That shouldn’t make any difference, `BinaryData fragmentElf = {}` still leaves the content of `codeSize` and `pCode` undefined, right?

Reply:
It doesn't really matter here, but yes, the `= {}` does make a difference. C++ is bonkers that way :)

Reply:
I should say, it is like C in that way: it's basically the degenerate form of a struct initialization with designated initializers.

Reply:
@Flakebi `= {}` will help you to zero all fields. Without `={}`, if a new field is uninitialized, it could take unexpected non-zero values. And all too often, structure tends to be extended by someone else besides the original author, which introduces latent issues :)

Reply:
Interesting, I didn’t know of that feature.